File: examples/zig/ec-unit/ECUnit.runar.zig
const runar = @import("runar");

pub const ECUnit = struct {
    pub const Contract = runar.SmartContract;

    pubKey: runar.ByteString,

    pub fn init(pubKey: runar.ByteString) ECUnit {
        return .{ .pubKey = pubKey };
    }

    pub fn testOps(self: *const ECUnit) void {
        _ = self;
        const g = runar.ecMulGen(1);
        runar.assert(runar.ecOnCurve(g));
        const neg = runar.ecNegate(g);
        runar.assert(runar.ecOnCurve(neg));
        const doubled = runar.ecMul(g, 2);
        runar.assert(runar.ecOnCurve(doubled));
        const sum = runar.ecAdd(g, g);
        runar.assert(runar.ecOnCurve(sum));
        const x = runar.ecPointX(g);
        const y = runar.ecPointY(g);
        const rebuilt = runar.ecMakePoint(x, y);
        runar.assert(runar.ecOnCurve(rebuilt));
        const compressed = runar.ecEncodeCompressed(g);
        runar.assert(runar.len(compressed) == 33);
        runar.assert(true);
    }
};

File: examples/go/ec-unit/ECUnit.runar.go
package contract

import runar "github.com/icellan/runar/packages/runar-go"

// ECUnit -- Unit-style exercises for the secp256k1 EC built-ins.
type ECUnit struct {
	runar.SmartContract
	PubKey runar.ByteString `runar:"readonly"`
}

// TestOps exercises ecMulGen, ecOnCurve, ecNegate, ecMul, ecAdd, ecPointX,
// ecPointY, ecMakePoint, and ecEncodeCompressed.
func (c *ECUnit) TestOps() {
	g := runar.EcMulGen(1)
	runar.Assert(runar.EcOnCurve(g))
	neg := runar.EcNegate(g)
	runar.Assert(runar.EcOnCurve(neg))
	doubled := runar.EcMul(g, 2)
	runar.Assert(runar.EcOnCurve(doubled))
	sum := runar.EcAdd(g, g)
	runar.Assert(runar.EcOnCurve(sum))
	x := runar.EcPointX(g)
	y := runar.EcPointY(g)
	rebuilt := runar.EcMakePoint(x, y)
	runar.Assert(runar.EcOnCurve(rebuilt))
	compressed := runar.EcEncodeCompressed(g)
	runar.Assert(runar.Len(compressed) == 33)
	runar.Assert(true)
}
